feat: add status subresource with last sync and generation tracking

Flydiverny opened the pull request:
Currently kubernetes-external-secrets does excessive polling of all external secrets and doesn't respect the polling intervall. Everytime the watch stream disconnects (happens somewhat regularly) all the secrets are regenerated. In our relatively small dev cluster this generates 22 million events to each of KMS, SM and GuardDuty which generates unnecessary billing, and a general waste of compute cycles for the sake of nothing.

This PR uses the status subresource to track generation of external secret, and the last sync time.
Will use these values to determine how to queue up the next poll, either instant or time remaining for next poll intervall to pass. This helps for setting long poll intervals as currently setting something like 24h and having the pod moved or watch stream disconnected would restart the interval from that time, instead of forcing unwanted polls.

Updated CRD to show fields when using kubectl:
![image](https://user-images.githubusercontent.com/326929/68081136-52462400-fe09-11e9-818b-416689a31f93.png)

closes #131 
fixes #117
fixes #153 
fixes #190
start of #185

Flydiverny commented:
@silasbw @jeffpearce Please have a look :D

jeffpearce commented:
Hi @Flydiverny. It seems like it's much simpler to just always upsert the secret when we start up. Do you have concerns around doing it that way?

Flydiverny commented:
@jeffpearce Cost would be a concern, and also it defeats the purpose of having a polling interval setting if its not respected.

As mentioned in #117 having a longer interval (which we do as well) the interval will currently (version 1.3.1) never happen or possibly happen in a totally different interval if either: the watch stream disconnects or the pod is restarted.

Say the watch stream doesn't seem to remain connected forever so if I then use a 24h interval and the watch stream only lasts 1h, I would never trigger the interval at 24h because every hour the 24h countdown is restarted as the poller restarts. I don't see any way around this without keeping the state somewhere (which in this PR is in the created secrets).

Further building onto this I think we could allow for specifying a polling interval per external secret as well. Could be useful for testing or in cases where some secrets rotate more often. Possibly even disabling polling for some and only update when external secret is updated.

silasbw commented:
Hi @Flydiverny, which cost are you thinking of?  The cost of polling AWS SM?

(Sorry for the delay, been on vacation)

chadlwilson commented:
Any chance of us getting a new release with better support for the longer poll intervals either with/via this PR or with the simpler fix already on master?

Flydiverny commented:
> Hi @Flydiverny, which cost are you thinking of? The cost of polling AWS SM?
> 
> (Sorry for the delay, been on vacation)

Yes this was referencing the costs of polling AWS SM, after running external-secrets for awhile now in our environments we can see we get unnecessarily expensive bills for the excessive amount of polling that is done.

Flydiverny commented:
Refactored to use status subresource instead.
Added generation tracking and updated CRD to show last sync and a status when using kubectl
![image](https://user-images.githubusercontent.com/326929/68081136-52462400-fe09-11e9-818b-416689a31f93.png)

Got some tests to fix if there's any interest in this.

Refactor will also got a bit of the way on #185 

See also #190 which points out the issue I'm trying to solve here :)

Flydiverny commented:
Tests fixed and stuff squashed and rebased to madness.

Flydiverny commented:
For anyone that wants to try this i pushed an image here https://hub.docker.com/r/flydiverny/kubernetes-external-secrets/tags

```sh
docker pull flydiverny/kubernetes-external-secrets:1.6.0-with.pr.133
```

Flydiverny commented:
@jeffpearce @silasbw @keweilu Please see updated description and review updated PR! :)

chadlwilson commented:
Thanks for your work here @Flydiverny 🎉

silasbw commented on lib/poller.js:
```diff
@@ -74,49 +94,68 @@ class Poller {
+      await kubeNamespace.secrets(this._name).put({ body: secretManifest })
     }
+
+    await this._updateStatus('SUCCESS')
```
nit: move this to `._updateStatus` to https://github.com/godaddy/kubernetes-external-secrets/pull/133/files#diff-ef923d397aa8d0a9b6bc6d401b5290d7R101.  Then the 'SUCCESS' and 'ERROR' status setting are in the same function. When I was reading this the first time I aw the 'ERROR' status and wondered why there wasn't a 'SUCCESS' status near by.

silasbw commented on lib/poller.js:
```diff
@@ -150,37 +189,66 @@ class Poller {
+      const lastPollNbr = Date.parse(lastSync) || 0
+      const now = Date.now()
+
+      if (lastPollNbr > now) {
```
In what cases would you expect this to be true?

Flydiverny replied:
Well never TBH, unless someone messes up their time zones or so :D I just ran into it when messing with the tests (intentionally setting bad dates). Could remove it

silasbw commented on lib/poller.js:
```diff
@@ -150,37 +189,66 @@ class Poller {
+        return this._setNextPoll(0)
+      }
+
+      const lastPollNbr = Date.parse(lastSync) || 0
```
I don't understand this variable name.  What's `Nbr`?

How about `lastPollDate` or `lastPollTime`?

Flydiverny replied:
Bad naming from first pass :tada: I'll rename it to time since it's ms

silasbw commented on lib/poller.js:
```diff
@@ -150,37 +189,66 @@ class Poller {
+        return this._setNextPoll(this._intervalMilliseconds)
       }
+
+      const nextPollIn = Math.max(lastPollNbr - (now - this._intervalMilliseconds), 0)
```
Can you add a variable to make this slightly easier to read?

```js
const elapsedTime = now - lastPollTime
const nextPollIn = Math.max(this._intervalMilliseconds - elaspedTime, 0)
```

silasbw commented on lib/poller.js:
```diff
@@ -150,37 +189,66 @@ class Poller {
-  async _checkForSecret () {
-    const kubeNamespace = this._kubeClient.api.v1.namespaces(this._namespace)
-
+  async _checkExternalSecretStatus () {
```
naming nit: how about `_scheduleNextPoll`?  I think that more accurately describes what this function does.

silasbw commented on lib/poller.js:
```diff
@@ -23,27 +23,47 @@ class Poller {
    * @param {string} namespace - Kubernetes namespace.
-   * @param {SecretDescriptor} secretDescriptor - Kubernetes secret descriptor.
+   * @param {Object} customResourceManifest - CRD manifest
+   * @param {Object} externalSecret - ExternalSecret custom resource object.
```
comment nit: "ExternalSecret manifest" instead of "ExternalSecret custom resource object".  When I see "custom resource object" i think CRD / Custom Resource Definition object.

Flydiverny replied:
Thanks it didn't feel right :smile:

silasbw commented on lib/daemon.js:
```diff
@@ -68,21 +49,16 @@ class Daemon {
+      throw new Error(`Poller for ${descriptor.id} already exists`)
+    }
+
+    const poller = this._pollerFactory.createPoller(descriptor)
```
I'm a little confused at the types here: `descriptor` is result of `_createPollerDescriptor`, but `.createPoller` takes an `ExternalSecret` manifest / object.  Is there an actual type problem, or confusion with variable naming?

Flydiverny replied:
Hmm well the full externalSecret now exists in in the descriptor, just destructuring externalSecret from the descriptor as nothing more is needed to create the poller, any suggestions to make it more clear?

Ie `this._pollerFactory.createPoller({ externalSecret: descriptor.externalSecret} )`

silasbw replied:
doh, i missed the destructuring in the `.createPoller`.  ignore me :)

Flydiverny commented on lib/daemon.js:
```diff
@@ -68,21 +49,16 @@ class Daemon {
+
+    if (this._pollers[descriptor.id]) {
+      throw new Error(`Poller for ${descriptor.id} already exists`)
+    }
```
So after running this for a few hours, this seems to happen occasionally, not really sure whats triggering it. 

Whenever an external secret is polled we will update the status for this external secret, this in turn should trigger the watch stream with a modified event. Which means we recreate the poller and schedule the next poll. As we track the generation with observedGeneration we won't end up in an endless loop as generation is not updated by status updates.

Flydiverny replied:
With change in bb3787c this now seems fine.

Switched back to returning in _upsertSecret here: https://github.com/godaddy/kubernetes-external-secrets/pull/133/commits/cf17824cde407c9bfed571b4bd72c20b1e9a3fe9
But this seemed to cause issues with errors getting caught in the wrong try catch. ie upserts failing. so added the await again https://github.com/godaddy/kubernetes-external-secrets/pull/133/commits/c8be7f389390a9e7de8046b2d28607f39c533a5a

Running test in our dev cluster again 🎉

silasbw commented on lib/poller.js:
```diff
@@ -150,37 +189,62 @@ class Poller {
       }
+
+      const lastPollTime = Date.parse(lastSync) || 0
+      const elapsedTime = Date.now() - lastPollTime
```
I didn't mean to imply we should remove it.  I like the defensive behavior of scheduling it immediately (and then hopefully getting a more sane lastPollTime value).  Maybe bring the code back and add a comment?

Flydiverny replied:
Ah yes, must have been tired last night. I was thinking this became a non issue after the changes but a negative elapsedTime here gives a funny next poll time 😅 I'll add it back.

keweilu commented on charts/kubernetes-external-secrets/templates/rbac.yaml:
```diff
@@ -25,6 +25,9 @@ rules:
   - apiGroups: ["kubernetes-client.io"]
     resources: ["externalsecrets"]
     verbs: ["get", "watch", "list"]
+  - apiGroups: ["kubernetes-client.io"]
```
can we also update the rbac in external-secrets.yml file https://github.com/godaddy/kubernetes-external-secrets/blob/master/external-secrets.yml#L18 with those permissions

Flydiverny replied:
Fixt !
